### PR TITLE
Move testing over to pytest and incorporate codecov

### PR DESCRIPTION
**Pull request by davidwyld**
This PR:
 - moves on from the tests being carried out in shell scripts and moves to using pytest
 - the tests take exactly the same format but are now written in python in `tests/BiRNNsent.py` `..doc.py` and `..word.py`
- various testing utility functions are moved to `test/utils.py` including the get test data functions and functions for extracting values from the testVals.csv and from the model outputs and predictions.
 - the tests are run one after the other with pytest, this is because:
    - if they're all in the same `.py` file, pytest runs them all in the same session,
    - this exposes a bug in which training runs that are run one after another in the same python session affect the outcome of subsequent training runs,
    - I've tried various attempts at working this out to no avail. I suspect that there is something going on in the keras session as a result of keras_wrapper or nmt-keras.
- ~Currently, there is a `with pytest.raises(SystemExit):` around `dq.train()`, because keras_wrapper's EarlyStopping callback exits with `exit(1)` which calls `SystemExit`, which would otherwise result in an error in pytest. I'm going to suggest that they get rid of this and use `return` instead.~ <- This now solved since https://github.com/MarcBS/multimodal_keras_wrapper/issues/224

### CODE REVIEW

**Comment by davidwyld** [comment]
@fredblain please review this PR so it can be merged.

File: tests/BiRNNdoc.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+	model = 'trained_models/' + model_name +'/epoch_' + epoch + '.h5'
+	dataset = 'trained_models/' + model_name + '/Dataset_' + task_name + '_srcmt.pkl'
+	save_path = 'saved_predictions/prediction_' + task_name + '/'
+	evalset = 'test'
+	dq.predict(model=model, dataset=dataset, save_path=save_path, evalset=evalset)
+	result = getPredictedVal(save_path + 'test.qe_metrics', metric)
+	testVal = getTestVal(backend, level, 'predict', task_name, metric)
+
+	assert (result - testVal)**2 < 1E-10 # check that output is within 1E-5 of the expected result
```
Comment (drj11) [inline tests/BiRNNdoc.py]: Now that we are in Python, we can program like a normal person and use

    assert abs(result - testVal) < 1E-5

(repeated elsewhere too)

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+    totalLines = 500 # total number of lines to take from example data
+    for f in os.listdir( cachePath ):
+        if f.endswith(".hter") or f.endswith(".pe") or f.endswith(".source") or f.endswith(".target"):
+            file_in = cachePath+f
+            file_out = dataDir+f
+            print('Copying first ' + str(totalLines) + ' lines of ' + file_in + ' to ' + file_out)
+            with open(file_in) as file:
+                lines = file.readlines()
+                lines = [l for i, l in enumerate(lines) if i <= totalLines-1]
```
Comment (drj11) [inline tests/utils.py]: `i < totalLines` neater than `i <= totalLines-1`. You've done that elsewhere, so this is presumably me failing to spot this on some previous occasion.

**Review by drj11 — APPROVED** [review]

File: tests/utils.py (hunk 2 above)
Comment (davidwyld) [reply tests/utils.py]: Yes, sorry, this must be a hangover from some previous foible.
Resolved by 421fb96

File: tests/BiRNNdoc.py (hunk 1 above)
Comment (davidwyld) [reply tests/BiRNNdoc.py]: Quite right!
Resolved by 785ce5b

**Pull request merged**